File: docs/manual/guide/cluster/code/docs/home/persistence/BlogCommand.java
package docs.home.persistence;

//#full-example
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.lightbend.lagom.javadsl.immutable.ImmutableStyle;
import com.lightbend.lagom.javadsl.persistence.PersistentEntity;
import com.lightbend.lagom.serialization.Jsonable;
import org.immutables.value.Value;
import akka.Done;

public interface BlogCommand extends Jsonable {

  //#AddPost
  @Value.Immutable
  @ImmutableStyle
  @JsonDeserialize(as = AddPost.class)
  interface AbstractAddPost
    extends BlogCommand, PersistentEntity.ReplyType<AddPostDone> {

    @Value.Parameter
    PostContent getContent();
  }
  //#AddPost

  @Value.Immutable
  @ImmutableStyle
  @JsonDeserialize(as = AddPostDone.class)
  interface AbstractAddPostDone extends Jsonable {
    @Value.Parameter
    String getPostId();
  }

  @Value.Immutable(singleton = true, builder = false)
  @ImmutableStyle
  @JsonDeserialize(as = GetPost.class)
  public abstract class AbstractGetPost implements BlogCommand, PersistentEntity.ReplyType<PostContent> {
    protected AbstractGetPost() {
    }
  }

  @Value.Immutable
  @ImmutableStyle
  @JsonDeserialize(as = ChangeBody.class)
  interface AbstractChangeBody extends BlogCommand, PersistentEntity.ReplyType<Done> {
    @Value.Parameter
    String getBody();
  }

  @Value.Immutable(singleton = true, builder = false)
  @ImmutableStyle
  @JsonDeserialize(as = Publish.class)
  public abstract class AbstractPublish implements BlogCommand, PersistentEntity.ReplyType<Done> {
    protected AbstractPublish() {
    }
  }

}
//#full-example

File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/BlogCommand.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

//#full-example
import com.lightbend.lagom.scaladsl.persistence.PersistentEntity.ReplyType
import akka.Done
import com.lightbend.lagom.scaladsl.playjson.JsonSerializer

sealed trait BlogCommand

object BlogCommand {
  import play.api.libs.json._
  import JsonSerializer.emptySingletonFormat

  implicit val postContentFormat = Json.format[PostContent]

  val serializers = Vector(
    JsonSerializer(Json.format[AddPost]),
    JsonSerializer(Json.format[AddPostDone]),
    JsonSerializer(emptySingletonFormat(GetPost)),
    JsonSerializer(Json.format[ChangeBody]),
    JsonSerializer(emptySingletonFormat(Publish))
  )
}

//#AddPost
final case class AddPost(content: PostContent) extends BlogCommand with ReplyType[AddPostDone]
//#AddPost

final case class AddPostDone(postId: String)

case object GetPost extends BlogCommand with ReplyType[PostContent]

final case class ChangeBody(body: String) extends BlogCommand with ReplyType[Done]

case object Publish extends BlogCommand with ReplyType[Done]
//#full-example
